lib/ssh_scan/rule_engine.rb

```diff
@@ -0,0 +1,94 @@
+module SSHScan
+  class RuleEngine
+    # Simple one rule matching.
+    def self.matchSingleRule(key, value, result)
+      answer = true
+      conditional = key.split("_")[-1]
+      attribute = key.chomp("_" + conditional).to_sym
+      rvalue = result[attribute]
+      if rvalue
+        answer &= case conditional
+          when "eq" then (rvalue == value)
+          when "neq" then (rvalue != value)
+          else true
+        end
+      end
+      if rvalue.is_a? Numeric
+        answer &= case conditional
+          when "gt" then (rvalue > value)
+          when "lt" then (rvalue < value)
+          when "gte" then (rvalue >= value)
+          when "lte" then (rvalue <= value)
+          else true
+        end
+      end
+      answer
+    end
+
+    # Match a simple ruleset condition using the stack based parser.
+    def self.matchRuleset(rules, ruleset, result)
+      mem = {}
+      rules.each do |rule_name, values|
+        mem[rule_name] = matchSingleRule(values[0], values[1], result)
+      end
+      return parseBooleanRule(ruleset, mem)
+    end
+
+    # Check if o is boolean or not. Apparently is_a?(Boolean) is not a thing.
+    def self.checkBool(o)
+      !!o == o
+    end
+
+    # Reduce boolean expression parser stack once.
+    def self.reduceStack!(stack)
+      if (stack[-2] == '!') && checkBool(stack[-1])
+        new_tokenval = !stack[-1]
+        2.times { stack.pop }
+        stack << new_tokenval
+      elsif (stack[-2] == '&&') && checkBool(stack[-3]) && checkBool(stack[-1])
+        new_tokenval = stack[-1] && stack[-3]
+        3.times { stack.pop }
+        stack << new_tokenval
+      elsif (stack[-2] == '||') && checkBool(stack[-3]) && checkBool(stack[-1])
+        new_tokenval = stack[-1] || stack[-3]
+        3.times { stack.pop }
+        stack << new_tokenval
+      else
+        return false
+      end
+      return true
+    end
```

Comment:
I'm not sure what this is doing. Maybe some more documentation around this functionality would help.